### ISSUE TITLE
KeepRight interface followup issues

### ISSUE THREAD

**Issue by bhousel** [openstreetmap/iD#5679]
Breaking out some of @tordans's feedback into a separate issue.  I think all the other ones from the comment are fixed already, such as fixing the layer order to make it easier to click on things, offsetting the markers slightly from the geometry they are about, and improving the color scheme and contrast of the markers, but let me know if I cut anything important.

> it took me quite a while to understand what the issue is related to, since I did not understand the "way 35557143"  (left panel) as the part that is most relevant for this. 

Yes, I'd like to make some of the embedded names nicer instead of showing entity ids.

> d. Is there a validation on the comment field? I feel like it should validate to at least 2 word for ignore cases, right? I dont want to test it since I dont know what will happen and the data are live, right?

KeepRight allows only a single anonymous comment.  You can change it or remove it, but there's not much else to know about commenting. I think I'm ok with just letting people play with the comment field, but open to ideas on ways to make it more clear for users.

> e. I image when I could actually fix the issue, I have the same problem as with https://github.com/openstreetmap/iD/issues/5170, I would like to manually sync my commit message with the get-right-issue. This is the same for HOT Task manager btw,  I type my change twice, once in the task manager, once in the commit message. Could the QA layer maybe pre-fill the commit-message with a hash-tag like "#keepright #keepright-issue-123" so there is some link?

Yes this is a good suggestion.  I'm wondering whether we should introduce a dedicated changeset tag to hold all the issues that users are closing?

> f. How do I get from one issue to the next one? Lets say I solved this one, how do I find the next one? 

There is parallel work going on where we are introducing an Issue Manager for both server-side and client-side validation issues.. The KeepRight issues will probably end up in here.

> g. I am looking for a way to share a specific keepright-issue on iD editor. I can share the Keepright-URL, which is provided in iD, but what I was hoping/looking for is: Once I select an KR-issue, the iD URL changes to reflect this issue. 

Yeah we should consider this - there are lots of other requests to use URL parameters to show different layers and focus on things in those layers (e.g. Mapillary/OpenStreetCam detections too) 

_Originally posted by @tordans in https://github.com/openstreetmap/iD/pull/5201#issuecomment-417834268_

**Comment by SilentSpike** [comment]
Regarding
>it took me quite a while to understand what the issue is related to, since I did not understand the "way 35557143" (left panel) as the part that is most relevant for this.

I envision something along the lines of how relation members are listed working well here - especially with the same highlighting to show the entities

**Comment by bhousel** [comment]
I added some code to replace the identifier with a friendlier name if possible. (This is not always possible because you might be viewing the markers at a lower zoom and the map data hasn't been loaded around there).  

It will also highlight the feature if you hover over the link, and zooming to the entity works a little better now.

<img width="856" alt="screenshot 2019-01-07 16 04 20" src="https://user-images.githubusercontent.com/38784/50793354-ec931500-1295-11e9-90bf-3271bb6e9b8e.png">

**Comment by SilentSpike** [comment]
Another suggestion:

The left panel should also have a link somewhere to the object the issue as actually attached to as it may not always be obvious which it is (currently it only links to additional entities listed in some descriptions).

It would also serve as a good way to zoom to the error when browsing them from a low zoom level.

**Comment by thomas-hervey** [comment]
> The left panel should also have a link somewhere to the object the issue as actually attached to as it may not always be obvious which it is (currently it only links to additional entities listed in some descriptions).

- [ ] @SilentSpike, I agree and this is an easy fix. This was in my original implementation but I think it may have been hidden when @bhousel worked on cleaner naming and descriptions. I think it should be re-enabled.

**Comment by bhousel** [comment]
> * @SilentSpike, I agree and this is an easy fix. This was in my original implementation but I think it may have been hidden when @bhousel worked on cleaner naming and descriptions in [2ec02f3](https://github.com/openstreetmap/iD/commit/2ec02f369f02626cd94c0a47d9ebb698c8fca431) and [9557c02](https://github.com/openstreetmap/iD/commit/9557c0204a2c024c1561450e799a7c1f0264fd99). I think it should be re-enabled.

I'm not sure that we get the geometry that the issue is attached to from KeepRight.. For example in this one, the original message was "This node is very close but not connected to way #11775974".  KeepRight doesn't give us the "this node" part.  (we could probably guess 🤔)

<img width="819" alt="screenshot 2019-01-09 23 11 06" src="https://user-images.githubusercontent.com/38784/50945704-f0749200-1463-11e9-8b05-735f0e86dd86.png">

**Comment by thomas-hervey** [comment]
@bhousel, if I understand correctly, we _do_ get the geometry that the issue is attached to. It's called _object_id_

Before the header was changed, this object id was visible and clickable. You can see examples in @SilentSpike's comment https://github.com/openstreetmap/iD/pull/5275#issuecomment-418100026
I'll have to dig up a commit I made to see where the object id gets used.

**Comment by bhousel** [comment]
> @bhousel, if I understand correctly, we _do_ get the geometry that the issue is attached to. It's called _object_id_

Nice!  Thanks @thomas-hervey..  Sorry if I removed it - I will find a way to put them back..

**Comment by SilentSpike** [comment]
Yeah an object ID and type is being [served in the json](https://github.com/openstreetmap/iD/blob/98c08a02efe30aab2482e1a0e89267ff88957a83/modules/services/keepRight.js#L328-L329), whether or not it links to the specific element for evey error type I don't know though

**Comment by bhousel** [comment]
> Yeah an object ID and type is being [served in the json](https://github.com/openstreetmap/iD/blob/98c08a02efe30aab2482e1a0e89267ff88957a83/modules/services/keepRight.js#L328-L329), whether or not it links to the specific element for evey error type I don't know though

Yep - I'm updating all the strings today to support adding links to the "this whatever" part of the message.  

In a few cases, I'm rewriting the strings to make this easier to do.   For example, one of the errors we get just says "missing maxspeed tag", so in code I'm changing that to "This highway is missing a maxspeed tag" (to give us something to actually add a link to).

**Comment by bhousel** [comment]
I redid all the error messages to make sure they contain links to the object.  This unfortunately means a lot more translation work, but the error messages are a lot better now!  

I also made a few other small improvements to the links.  If the user clicks a link to an entity that is not yet loaded into the graph (maybe they are zoomed out too far), we now load the entity from OSM before entering select mode.

**Comment by bhousel** [comment]
> I notice the URL errors (411, 412, 413) haven't been given this treatment, was that intentional?

oh you're right!  I forgot those ones..  I will add them.

**Comment by Mahabarata** [comment]
Hi,

I don't want to open a new issue, maybe this is the good place for a suggestion about Keepright.
With the last release, I really love the add of Keepright in ID : very easy to use, it is a very good improvment of ID, you have made a very good job, thanks a lot.
But when I check the box then close my browser, when I come back Keepright is unchecked : the first time I was happy because I did not see any more "bugs" in OSM in the area where I was working and then I saw the unchecked box.
It will be great if the checked box could stay.
But in fact no : I think that Keepright is a tool everyone should use (or at least that everyone should see the bugs shown by this tool) so I think it will be better to have no choice : no box to check but Keepright always there.
Because I think that a lot of people have not seen this box and don't know of your great improvment.
Why did you hide it, if you know what I mean ?

Best regards

**Comment by quincylvania** [comment]
@Mahabarata Good suggestion about persisting the KeepRight visibility setting. We should probably persist more of the other settings as well, like photo overlay and OSM note visibility.

This drawback to always showing KeepRight issues is that it could be overwhelming to new users and cumbersome for power-mappers. However, the next release of iD will include #5830 with numerous improvements to iD-native validation, such as flagging KeepRight-like issues as you edit.

**Comment by Mahabarata** [comment]
I understand what you mean, I just find this kind of result for the first time :
![2019-02-15_155056](https://user-images.githubusercontent.com/3121119/52884059-eae12a80-313a-11e9-8ecb-b10b27aba044.jpg)
So forget my suggestion of always showing KeepRight, sorry for this bad idea.
Best regards

**Comment by matkoniecz** [comment]
keepright was removed in https://github.com/openstreetmap/iD/commit/7555ff8291f1f9b78321a9bbb687db54e5e77899

**Comment by Mahabarata** [comment]
Because keepright is out to date since 2024-04, see bottom left of https://keepright.at/report_map.php?lang=en&ch30=1&ch40=1&ch50=1&ch70=1&ch90=1&ch100=1&ch110=1&ch120=1&ch130=1&ch150=1&ch160=1&ch170=1&ch180=1&ch191=1&ch192=1&ch193=1&ch194=1&ch195=1&ch196=1&ch197=1&ch198=1&ch201=1&ch202=1&ch203=1&ch204=1&ch205=1&ch206=1&ch207=1&ch208=1&ch210=1&ch220=1&ch231=1&ch232=1&ch270=1&ch281=1&ch282=1&ch283=1&ch284=1&ch285=1&ch291=1&ch292=1&ch293=1&ch294=1&ch295=1&ch296=1&ch297=1&ch298=1&ch311=1&ch312=1&ch313=1&ch320=1&ch350=1&ch370=1&ch380=1&ch401=1&ch402=1&number_of_tristate_checkboxes=7&highlight_error_id=0&highlight_schema=0&lat=51.76056799168763&lon=-1.2095696803163898&zoom=14&show_ign=1&show_tmpign=1
